lib/graphql/object_type.rb

```diff
@@ -97,25 +106,37 @@ def normalize_interfaces(ifaces)
     end
 
     def interface_fields
-      load_interfaces
-      @clean_inherited_fields
-    end
-
-    def load_interfaces
-      @clean_interfaces ||= begin
-        ensure_defined
-        clean_ifaces = normalize_interfaces(@dirty_interfaces)
-        clean_inherited_ifaces = normalize_interfaces(@dirty_inherited_interfaces)
-        inherited_fields = {}
-        clean_inherited_ifaces.each do |iface|
-          # This will be found later in schema validation:
+      if @clean_inherited_fields
+        @clean_inherited_fields
+      else
+        _ifaces, inherited_ifaces = load_interfaces(nil)
+        @clean_inherited_fields = {}
+        inherited_ifaces.each do |iface|
           if iface.is_a?(GraphQL::InterfaceType)
-            inherited_fields.merge!(iface.fields)
+            @clean_inherited_fields.merge!(iface.fields)
           end
         end
-        @clean_inherited_fields = inherited_fields
-        clean_inherited_ifaces + clean_ifaces
+        @clean_inherited_fields
+      end
+    end
+
+    def load_interfaces(ctx = GraphQL::Query::NullContext)
+      ensure_defined
+      clean_ifaces = []
+      clean_inherited_ifaces = []
+      @interface_type_memberships.each do |type_membership|
+        if ctx.nil? || type_membership.visible?(ctx)
```

Comment:
`ctx` defaults to `GraphQL::Query::NullContext` so I believe `ctx.nil?` isn't the correct option here?

Reply:
I need _some_ behavior to return all interfaces, without filtering. The library does that in a few places. I've chosen `interfaces(nil)`, where passing `nil` as context is the signal to return _everything_. In that case, `ctx` is `nil`. 

I'm not sold on this approach; another approach (which is taken in `Schema::Object`) is to expose `interface_type_memberships`, and let callers who need unfiltered access use _that_ instead. 

But anyways, that's why `nil` is handled explicitly here.

Reply:
👍 Makes sense that we need that case.

And also makes sense to make it a more explicit flag, however when I first was building this I was thinking that a NullContext was going to be that indicator if we needed to return everything, but perhaps NullContext is too widely used to be explicty return all?

Is there a case where we want to pass in a sample context of `{ return_all: true }` to be even more explicit?

Reply:
> a NullContext was going to be that indicator

That's a good idea. I just refactored it to remove the `.interfaces(nil)` use case, but I think you're right about using `NullContext` for it, and supporting `SomeObject.interfaces` returning _everything_.

The thing is, even if we fall back to `NullContext`, we have to hardcode a check for it, because passing `NullContext` _won't_ make `.visible?` checks pass. 

Anyhow, thanks for sharing your original intention here. Let me take another look.